[Lim Ji Wei] iP

Jiwei99 opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke is a Personal Assistant Chatbot that helps you keep track of your tasks!

## Benefits
Duke is
* text-based
* easy to learn
* ~~FAST~~ **_SUPER_ FAST** to use

## Features
- [x] Add Tasks: todo, deadline and event
- [x] List tasks
- [x] Filter tasks by keyword
- [x] Mark tasks as done
- [x] Delete task
- [ ] GUI (coming soon)

## Download
To use Duke, all you need to do is
1. download it from [here](https://github.com/Jiwei99/ip/releases/download/A-Jar/ip.jar)
2. double click the `.jar` file
3. add your tasks
4. let Duke manage your tasks for you! 😀

## Developer Notes
If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```java
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.run();
}
```

Commit:
Add Gradle support

Jiwei99 committed:
Implement a skeletal version of Duke

Jiwei99 committed:
Implement ability to store text

Jiwei99 committed:
Add the ability to mark tasks as done.

Jiwei99 committed:
Add support for tracking three types of tasks

Jiwei99 committed:
Test using the I/O redirection technique, minor fixes

Jiwei99 committed:
Implement error handling

Jiwei99 committed:
Update automated text UI testing

Jiwei99 committed:
Implement delete task

Jiwei99 committed:
Add exception handling for missing index for done and delete

Jiwei99 committed:
Use enums

Jiwei99 committed:
Add saving to disk

Duke can now write and read tasks from disk

Jiwei99 committed:
Add support for dates and times

Duke can now understand dates and times

Jiwei99 committed:
Merge branch 'branch-Level-8'

Jiwei99 committed:
Fix SaveHelper to support LocalDate and LocalDateTime

SaveHelper can now parse LocalDate and LocalDateTime for read and write

Jiwei99 committed:
Refactor code to extract out classes

Extract Parser, TaskList and Ui classes from Duke class

Jiwei99 committed:
Add packaging

Put files into their respective packages

Jiwei99 committed:
Add JUnit Tests

Add JUnit Tests for Todo, Deadline and Event

Jiwei99 committed:
Create a JAR File

Create a JAR File for Duke

Jiwei99 committed:
Add JavaDocs

Add JavaDocs for all the Class files and the necessary methods.

Jiwei99 committed:
Update coding standard

Refactor code to follow coding standard.

Jiwei99 committed:
Add find function

Add function to allow users to find tasks using keywords

Jiwei99 committed:
Merge branch 'branch-A-CodingStandard'

Jiwei99 committed:
Merge branch 'branch-A-JavaDoc'

Jiwei99 committed:
Merge branch 'branch-Level-9'

Jiwei99 committed:
Add JavaDoc for find function

Add JavaDocs for functions in relation to the implementation of the find functionality

BenjaminTan99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Represents a Deadline Task.
+ */
+public class Deadline extends Task{
```
Leave a space between Task and {

BenjaminTan99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Class constructor specifying the task and the deadline.
+     * @param task Task to be done.
```
Leave a line between all methods' descriptions and parameter sections (where applicable) as according to the Java Coding Standard:

"Empty line between description and parameter section"

Applies for all subsequent methods. Will use (JD) to denote subsequent similar issues below.

BenjaminTan99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Class constructor specifying the task, whether the task was completed and the deadline.
+     * @param task Task to be done.
```
(JD) as above.

BenjaminTan99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+     * @param date Deadline by which the task needs to be done.
+     */
+    public Deadline(String task, boolean completed, LocalDate date) {
+        super(task, completed,"D");
```
spacing between , and "D" should be , "D")

BenjaminTan99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Returns the dateline of the task in a pre-defined format.
+     * @return Deadline of task.
```
(JD) as above

BenjaminTan99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class Duke {
+
+    private final TaskList tasks;
```
Preferably have some class members documentation for Duke since Duke is your main program.

BenjaminTan99 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Represents an Event task.
+ */
+public class Event extends Task{
```
same issue as mentioned earlier with regards to spacing:  Task {

BenjaminTan99 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Class constructor specifying the task and the event date and time.
+     * @param task Event to attend.
```
(JD) as mentioned before

BenjaminTan99 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Class constructor specifying the task and the event date and time
+     * @param task Event to attend.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Returns the date and time of the event in a pre-defined format.
+     * @return Date and time of Event.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Parses and executes the command that the user input.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+
+    /**
+     * Terminates the Duke Personal Assistant Chatbot.
+     * @param ui UI object to render output to user.
```
(JD) as mentioned earier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+
+    /**
+     * Clears all tasks in the TaskList.
+     * @param tasks TaskList object that contains the tasks.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+
+    /**
+     * Adds a ToDo task to the TaskList.
+     * @param tasks TaskList object that contains the tasks.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Adds a Deadline task to the TaskList.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Adds an Event task to the TaskList.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Marks a task as done.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+
+    /**
+     * Deletes a task.
+     * @param tasks TaskList object that contains the tasks.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+
+    /**
+     * Finds tasks that contains keyword.
+     * @param tasks TaskList object that contains the tasks.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+package duke;
+
+import java.io.IOException;
+import java.nio.file.Files;
```
Ordering of import statements is inconsistent. Try to order it according to alphabetical order and segregate the types of imports as well for clarity. You can refer to the Java Coding Standard Website.

BenjaminTan99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * Reads the TaskList from local storage.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * Writes the TaskList into local storage.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,83 @@
+    private String taskType;
+
+    /**
+     * Class constructor specifying the task description.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * Class constructor specifying the task description and the task type.
```
(JD) as mentioned earlier.

BenjaminTan99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * Class constructor specifying the task description, if the task was completed and the task type.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,83 @@
+
+    /**
+     * Returns a marker to represent if the task has been completed.
+     * @return String marker denoting if task has been completed.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * Checks if task description contains keyword.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * Parses the task to the appropriate format for local storage.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+
+    /**
+     * Marks a task as complete
+     * @param index Index of the task in the TaskList
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+
+    /**
+     * Returns a String indicating the number of tasks in the TaskList
+     * @return String representation of the number of tasks
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    /**
+     * Finds tasks that contains the keyword.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,34 @@
+public class ToDo extends Task{
+
+    /**
+     * Class constructor specifying the task description.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    /**
+     * Class constructor specifying the task description and if the task has been completed.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Prints a list of all the tasks
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Prints message indicating task has been added successfully.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Prints message indicating that task has been marked as done.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Prints message indicating that task has been completed.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,178 @@
+
+    /**
+     * Prints tasks containing specified keyword.
+     * @param tasks TaskList containing filtered tasks.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Prints the appropriate error handling message.
```
(JD) as mentioned earlier

BenjaminTan99 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Represents exceptions thrown by Duke Personal Assistant Chatbot.
+ */
+public class DukeException extends Exception{
```
leave a space between Exception and {

BenjaminTan99 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Represents exception where task description is not found.
+ */
+class DescriptionNotFoundException extends DukeException{
```
leave a space between DukeException and {

BenjaminTan99 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Represents exception where ToDo task description is not found.
+ */
+class ToDoDescriptionNotFoundException extends DescriptionNotFoundException{
```
Leave a space between DescriptionNotFoundException and {

BenjaminTan99 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Represents exception where Deadline task description is not found.
+ */
+class DeadlineDescriptionNotFoundException extends DescriptionNotFoundException{
```
Leave a space between DescriptionNotFoundException and {

BenjaminTan99 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Represents exception where Event task description is not found.
+ */
+class EventDescriptionNotFoundException extends DescriptionNotFoundException{
```
Leave a space between DescriptionNotFoundException and {

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+     * @throws DeadlineNotFoundException If user did not enter a deadline.
+     */
+    public void addDeadline(TaskList tasks, Ui ui, String input) throws DeadlineDescriptionNotFoundException,
+            DeadlineNotFoundException{
```
Leave a space between DeadlineNotFoundException and {

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+     * @param index Index of the task to be marked as done.
+     * @throws TaskIndexOutOfBoundException If no task bears the index.
+     */
+    public void markDone(TaskList tasks, Ui ui, int index) throws TaskIndexOutOfBoundException{
```
Leave a space between TaskIndexOutOfBoundException and {

BenjaminTan99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+     * @param index Index of the task to be deleted.
+     * @throws TaskIndexOutOfBoundException If no task bears the index.
+     */
+    public void delete(TaskList tasks, Ui ui, int index) throws TaskIndexOutOfBoundException{
```
Leave a space between TaskIndexOutofBoundException and {

BenjaminTan99 commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Represents a ToDo task.
+ */
+public class ToDo extends Task{
```
Leave a space between Task and {

BenjaminTan99 reviewed (commented):
Besides a little bit of JavaDocs changes required, it LGTM. Code is quite simple to read and understand. 👍

Jiwei99 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

Jiwei99 committed:
Add Gradle support

Jiwei99 committed:
Merge branch 'branch-A-Gradle'

zzybluebell commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,52 @@
+        Deadline deadline = new Deadline("testGetDate", true, LocalDate.parse("29-08-2021", DateTimeFormatter.ofPattern("dd-MM-yyyy")));
+        assertEquals("29 Aug 2021", deadline.getDate());
+    }
+}
```
Overall, looks very nice, awesome and I have no much more advice

zzybluebell commented on src/test/java/duke/EventTest.java:
```diff
@@ -0,0 +1,52 @@
+    public void testGetTime() {
+        Event event = new Event("testGetTime", true, LocalDateTime.parse("29-08-2021 0000", DateTimeFormatter.ofPattern("dd-MM-yyyy HHmm")));
+        assertEquals("29 Aug 2021 00:00", event.getTime());
+    }
```
Very sufficient test Junit Test case

Jiwei99 committed:
Implement GUI for user

Jiwei99 committed:
Implement GUI for user

Jiwei99 committed:
Implement command support for GUI

Duke commands can now be used with GUI

Jiwei99 committed:
Merge branch 'branch-Level-10'

Jiwei99 committed:
Add Java assertions

Duke does not utilise assertions to verify program state during runtime.

Assertions should be used to allow for checking for bugs.

Let's add assertions into the code where appropriate, such as when
reading from storage file, and when processing commands.

The use of assertions will allow checking for possible bugs in the code.

Jiwei99 committed:
Add more assertions

Jiwei99 committed:
Merge pull request #1 from Jiwei99/branch-A-Assertions

Add more assertions to Duke

Jiwei99 committed:
Improve code quality

Duke contains some code that does not follow good code quality
practices.

Code that do not follow good code quality practice will undermine
understandability and readibility of the code

Let's amend the code to follow the good code quality practices

Improving code quality will increase understandability and readability
of code.

Jiwei99 committed:
Merge pull request #3 from Jiwei99/branch-A-CodeQuality

Improve code quality

Jiwei99 committed:
Add priority tagging for tasks

All the tasks that the user input have equal priority.

Different tasks may have different priority and urgency and users may
want to complete some tasks before others.

Let's include a function to allow users to tag each task with a priority
ranking from 1 to 3 to indicate their urgency and importance level.

Adding priority tagging is preferble because users can now easily
decide which tasks to complete first.

Jiwei99 committed:
Merge pull request #4 from Jiwei99/branch-BCD-Extension

Add priority tagging for tasks

Jiwei99 committed:
Updates JavaDocs for Duke.

Jiwei99 committed:
Updates JavaDocs for Duke

Jiwei99 committed:
Improves on code quality

Jiwei99 committed:
Updates GUI to make it more user friendly

Jiwei99 committed:
Updates README.md

Jiwei99 committed:
Set theme jekyll-theme-architect

Jiwei99 committed:
Adds checkstyle to gradle

Jiwei99 committed:
Merge branch 'branch-A-CheckStyle'

Jiwei99 committed:
Merge branch 'master' of github.com:Jiwei99/ip

Jiwei99 committed:
Fix issue where full task list does not show.
